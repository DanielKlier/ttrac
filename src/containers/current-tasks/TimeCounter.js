import React from 'react';
import PropTypes from 'prop-types';
import {Label} from 'react-bootstrap';
import moment from 'moment';
import 'moment-duration-format';

export default class TimeCounter extends React.Component {

    // noinspection JSUnusedGlobalSymbols
    static PropTypes = {
        startDate: PropTypes.number.isRequired
    };

    constructor(props) {
        super(props);
        this.state    = {
            elapsedTime: 0
        };
        this.interval = 0;
    }

    componentWillMount() {
        this.interval = setInterval(() => this.tick(), 1000);
    }

    componentWillUnmount() {
        clearInterval(this.interval);
    }

    render() {
        return (
            <span className="task-list-hours">
                <Label bsStyle="info">
                    {
                        moment.duration(this.state.elapsedTime).format(
                            (this.state.elapsedTime > 60 * 60 * 1000)
                                ? 'h[h] m[m]'
                                : 'm[m] s[s]'
                        )
                    }
              </Label>
            </span>
        );
    }

    tick() {
        this.setState({
            ...this.state,
            elapsedTime: Date.now() - this.props.startDate
        });
    }
}
