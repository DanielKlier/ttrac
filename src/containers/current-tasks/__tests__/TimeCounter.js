import React from 'react';
import renderer from 'react-test-renderer';
import TimeCounter from '../TimeCounter';

jest.useFakeTimers();

const baseDate = new Date('2017-03-09T18:00:00').valueOf();

test('TimeCounter renders correctly', () => {

    Date.now = jest.fn(() => baseDate);

    const component = renderer.create(<TimeCounter startDate={baseDate}/>);

    expect(component.toJSON()).toMatchSnapshot();

    Date.now = jest.fn(() => baseDate + 1000);

    jest.runTimersToTime(1000);

    //component.update();

    expect(component.toJSON()).toMatchSnapshot();
});
