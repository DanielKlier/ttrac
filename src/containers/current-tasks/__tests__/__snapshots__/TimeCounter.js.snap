// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeCounter renders correctly 1`] = `
<span
  className="task-list-hours"
>
  <span
    className="label label-warning"
  >
    0s
  </span>
</span>
`;

exports[`TimeCounter renders correctly 2`] = `
<span
  className="task-list-hours"
>
  <span
    className="label label-warning"
  >
    1s
  </span>
</span>
`;
